File: examples/typescript/mongodb/mongodb_vfs.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { setServers } from 'node:dns'
import dotenv from 'dotenv'
import { MongoDBResource, MountMode, Workspace } from '@struktoai/mirage-node'

setServers(['8.8.8.8', '1.1.1.1'])
dotenv.config({ path: '.env.development' })

const uri = process.env.MONGODB_URI
if (uri === undefined) {
  console.error('MONGODB_URI missing in .env.development')
  process.exit(1)
}

const DEC = new TextDecoder()

async function dump(ws: Workspace, label: string, cmd: string): Promise<void> {
  console.log(`\n--- ${label} ---`)
  const r = await ws.execute(cmd)
  if (r.exitCode !== 0) {
    console.log(`(exit=${String(r.exitCode)}) ${DEC.decode(r.stderr)}`)
    return
  }
  process.stdout.write(DEC.decode(r.stdout))
  if (!DEC.decode(r.stdout).endsWith('\n')) process.stdout.write('\n')
}

async function main(): Promise<void> {
  const resource = new MongoDBResource({
    uri,
    defaultDocLimit: 50,
  })
  const ws = new Workspace({ '/mongodb/': resource }, { mode: MountMode.READ })

  try {
    console.log('=== VFS MODE: shell pipelines transparently read MongoDB ===')

    await dump(ws, 'ls /mongodb — databases', 'ls /mongodb')

    const dbsOut = await ws.execute('ls /mongodb')
    const dbs = DEC.decode(dbsOut.stdout).split('\n').filter((s) => s.length > 0)
    if (dbs.length === 0) {
      console.log('\nno databases')
      return
    }
    const target = dbs[0]!

    await dump(ws, `ls /mongodb/${target}`, `ls /mongodb/${target} | head -n 5`)

    const colsOut = await ws.execute(`ls /mongodb/${target}`)
    const cols = DEC.decode(colsOut.stdout).split('\n').filter((s) => s.endsWith('.jsonl'))
    if (cols.length === 0) {
      console.log(`\nno collections in ${target}`)
      return
    }
    const path = `/mongodb/${target}/${cols[0]!}`

    await dump(ws, `head -n 1 ${path}`, `head -n 1 ${path}`)
    await dump(ws, `wc -l ${path}`, `wc -l ${path}`)
    await dump(ws, `cat ${path} | wc -l`, `cat ${path} | wc -l`)
    await dump(ws, `jq -s ".[0]" ${path}`, `jq -s ".[0]" ${path}`)
  } finally {
    await ws.close()
    await resource.close()
  }
}

await main()

File: examples/python/mongodb/mongodb_vfs.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import json
import os
import sys

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.mongodb import MongoDBConfig, MongoDBResource

load_dotenv(".env.development")

config = MongoDBConfig(uri=os.environ["MONGODB_URI"])
resource = MongoDBResource(config=config)


async def main():
    with Workspace({"/mongodb/": resource}, mode=MountMode.READ) as ws:
        vos = sys.modules["os"]
        print("=== VFS MODE: open() reads from MongoDB ===\n")

        print("--- os.listdir() databases ---")
        databases = vos.listdir("/mongodb")
        for db in databases:
            print(f"  {db}")

        if not databases:
            print("no databases found")
            return

        db = "sample_mflix"
        if db not in [d.split("/")[-1] for d in databases]:
            db = databases[0].split("/")[-1]

        print(f"\n--- os.listdir() {db} collections ---")
        collections = vos.listdir(f"/mongodb/{db}")
        for col in collections:
            print(f"  {col}")

        if not collections:
            print("no collections found")
            return

        target = None
        for col in collections:
            if "movies" in col:
                target = col
                break
        if not target:
            target = collections[0]

        path = f"/mongodb/{db}/{target}"
        print(f"\n--- open() + read {target} ---")
        with open(path) as f:
            content = f.read()
        lines = [ln for ln in content.strip().split("\n") if ln.strip()]
        print(f"  documents: {len(lines)}")
        for line in lines[:3]:
            try:
                doc = json.loads(line)
                title = doc.get("title", doc.get("name", "?"))
                doc_id = doc.get("_id", "?")
                print(f"  [{doc_id}] {title}")
            except json.JSONDecodeError:
                print(f"  {line[:80]}")

        print("\n--- session observer ---")
        day_folders = vos.listdir("/.sessions")
        log_entries = vos.listdir(day_folders[0]) if day_folders else []
        for e in log_entries:
            print(f"  {e}")
        if log_entries:
            with open(log_entries[0]) as f:
                for i, line in enumerate(f):
                    if i >= 3:
                        break
                    print(f"  [{i}] {line.strip()[:120]}")

        records = ws.ops.records
        total = sum(r.bytes for r in records)
        print(f"\nStats: {len(records)} ops, "
              f"{total} bytes transferred")


asyncio.run(main())
